mongosql/src/algebrizer/definitions.rs

```diff
@@ -475,14 +477,61 @@ impl<'a> Algebrizer<'a> {
         }
 
         // Build the Project Stage using the source and built expression.
-        let stage = mir::Stage::Project(mir::Project {
+        let project_stage = mir::Stage::Project(mir::Project {
             source: Box::new(source),
-            expression,
+            expression: expression.clone(),
             is_add_fields,
             cache: SchemaCache::new(),
         });
-        stage.schema(&self.schema_inference_state())?;
-        Ok(stage)
+        project_stage.schema(&self.schema_inference_state())?;
+        Ok(if is_distinct {
+            let group_stage = Self::build_distinct_values_group_stage(
+                &expression,
+                project_stage,
+                self.scope_level,
+            );
+            group_stage.schema(&self.schema_inference_state())?;
+            group_stage
+        } else {
+            project_stage
+        })
+    }
+
+    fn build_distinct_values_group_stage(
+        project_tuple: &BindingTuple<Expression>,
+        source_stage: mir::Stage,
+        scope_level: u16,
+    ) -> mir::Stage {
+        let mut group_keys: Vec<OptionallyAliasedExpr> = Vec::new();
+
+        for (key, expr) in project_tuple.iter() {
+            if key.datasource == DatasourceName::Bottom && key.scope == scope_level {
+                if let Expression::Document(DocumentExpr { document }) = expr {
+                    for (field_name, field_expr) in document.iter() {
+                        let alias = field_name.to_string();
+                        let group_key_expr = Expression::FieldAccess(FieldAccess {
+                            expr: Box::new(Expression::Reference(ReferenceExpr {
+                                key: key.clone(),
+                            })),
+                            field: field_name.to_string(),
+                            is_nullable: field_expr.is_nullable(),
+                        });
+                        let group_key = OptionallyAliasedExpr::Aliased(AliasedExpr {
+                            alias: alias.clone(),
+                            expr: group_key_expr,
+                        });
+                        group_keys.push(group_key);
```

Comment:
We shouldn't need to depend on the `expr` being a document and we shouldn't need to group by the individual fields. Again, consider my example above, `SELECT DISTINCT VALUES foo.*, bar.* FROM foo JOIN bar`. That should translate roughly to something like:
```
[
  { $project: { _id: 0, foo: "$$ROOT" }},
  { $join: { collection: "bar", join_type: "inner", pipeline: [{ $project: { _id: 0, bar: "$$ROOT" } }] } },
  { $project: { _id: 0, foo: "$foo", bar: "$bar" } },
  { $group: { _id: {foo: "$foo", bar: "$bar" } } },
  { $project: { _id: 0, foo: "$_id.foo", bar: "$_id.bar" } }
]
```
I'm not sure this code will do that since it depends on finding `Bottom` and the `expr` being a `Document`.

Generally, I believe the group keys should just be the datasources themselves!

Reply:
I made the changes that should address this.  There are still some areas I'm not completely sure about such as adding the `BOTTOM_ALIAS`, and how to set is_nullable.  I'll look into it a bit more but might have to ask you for help tomorrow.